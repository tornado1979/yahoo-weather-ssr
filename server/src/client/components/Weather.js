import React, { Component } from 'react'
import { connect } from 'react-redux'
import propTypes from 'prop-types'
import { fetchData } from '../actions'
import {
  getWeather,
  getCities,
} from '../selectors'
import { Forecast } from './Forecast'
import Cities from './Cities'

class Weather extends Component {
  constructor(props) {
    super(props)
    this.state = {
      activeCity: 'San Diego, CA',
      indx: 1,
    }
  }

  componentDidMount() {
    const {
      fetchData: getData,
    } = this.props

    getData()

    // set 5secs interval between the cities weather information
    const { cities } = this.props

    setInterval(() => {
      this.setState({
        indx: this.state.indx + 1, // eslint-disable-line
        activeCity: cities[this.state.indx%cities.length],// eslint-disable-line
      })
      getData(this.state.activeCity)// eslint-disable-line
    }, 5000)
  }

  // render forecast days
  renderForecast(forecastArray) {
    return forecastArray.map(day => <li key={day.date}><Forecast forecast={day} /></li>)
  }

  // render whole weather component
  renderWeather() {
    const {
      weather,
      cities,
    } = this.props

    const {
      activeCity,
    } = this.state

    if (Object.keys(weather).length) {
      const {
        item,
        location,
        units,
      } = weather

      const forecast = this.renderForecast(item.forecast)

      return (
        [
          <div className="current" key="1">
            <div className="center">
              <div className="city">{location.city}</div>
              <div className="text">{item.condition.text}</div>
            </div>
            <div className="temp">
              {item.condition.temp}
            </div>
            <div className="unit">{units.temperature}</div>
          </div>,
          <ul className="forecast-list" key="2">
            {forecast}
          </ul>,
          <div className="bottom" key="3">
            {weather.title}
            <span style={{ color: '#fff', fontWeight: 600, padding: '0 0.5rem' }}>|</span>
            {weather.lastBuildDate}
          </div>,
          <div className="cities" key="4">
            <Cities activeCity={activeCity} cities={cities} />
          </div>,
        ]
      )
    }

    // in case something is wrong
    return <div>No Data. Something gone wrong :(</div>
  }

  render() {
    const {
      weather,
    } = this.props

    const dataLoaded = !!Object.keys(weather).length

    return (
      <div className="weather-container fade-in">
        {dataLoaded && this.renderWeather()}
        {!dataLoaded && <div className="loader">Data is loading...</div>}
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    cities: getCities(state),
    weather: getWeather(state),
  }
}

const loadData = (store) => {
  return store.dispatch(fetchData())
}

Weather.propTypes = {
  cities: propTypes.arrayOf(propTypes.oneOfType([propTypes.string])).isRequired,
  fetchData: propTypes.func.isRequired,
  weather: propTypes.shape().isRequired,
}

export { loadData }

export default connect(mapStateToProps, { fetchData })(Weather)
