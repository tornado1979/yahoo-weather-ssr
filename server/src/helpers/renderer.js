import React from 'react'
import { renderToString } from 'react-dom/server'
import { StaticRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { renderRoutes } from 'react-router-config'
import serialize from 'serialize-javascript'

import Routes from '../client/Routes'

export default (req, store, context) => {
  const content = renderToString(
    <Provider store={store}>
      <StaticRouter context={context} location={req.path}>
        <div>{renderRoutes(Routes)}</div>
      </StaticRouter>
    </Provider>,
  )
  return `
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="UTF-8">
      <title>Server Side render application</title>
    </head>
   <body>
     <div id="root">${content}</div>
     <script>
       window.INITIAL_STATE = ${serialize(store.getState())}
     </script>
     <script src="bundle.js"></script>
   </body>
  </html>
  `
}
