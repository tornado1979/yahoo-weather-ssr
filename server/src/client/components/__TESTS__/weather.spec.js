import React from 'react'
import renderer from 'react-test-renderer'

import { Weather } from '../Weather'

describe('Component Weather', () => {
  it('renders correctly', () => {
    const addCity = () => console.log('add city')
    const cities = ['Athens', 'Thessaloniki']
    const fetchData = () => console.log('fetch data')
    const isFetching = false
    const weather = {}
    const error = {}

    const tree = renderer
      .create(<Weather
        addCity={addCity}
        cities={cities}
        fetchData={fetchData}
        isFetching={isFetching}
        weather={weather}
        error={error}
      />)
    expect(tree).toMatchSnapshot()
  })
})
