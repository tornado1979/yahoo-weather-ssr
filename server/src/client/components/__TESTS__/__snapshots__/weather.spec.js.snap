// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Weather renders correctly 1`] = `
<div
  className="weather-container fade-in"
>
  <div
    className="loader"
  >
    No data for that City :(
  </div>
</div>
`;
